ipc/nodepart/arwenDriver.go

```diff
@@ -0,0 +1,308 @@
+
+	driver.resetPipeStreams()
+
+	arwenPath, err := driver.getArwenPath()
+	if err != nil {
+		return err
+	}
+
+	arguments, err := common.PrepareArguments(driver.vmType, driver.blockGasLimit, driver.gasSchedule, logger.LogDebug)
+	if err != nil {
+		return err
+	}
+
+	driver.command = exec.Command(arwenPath, arguments...)
+	driver.command.ExtraFiles = []*os.File{driver.arwenInputRead, driver.arwenOutputWrite, driver.arwenLogWrite}
+
+	arwenStdout, err := driver.command.StdoutPipe()
+	if err != nil {
+		return err
+	}
+
+	arwenStderr, err := driver.command.StderrPipe()
+	if err != nil {
+		return err
+	}
+
+	err = driver.command.Start()
+	if err != nil {
+		return err
+	}
+
+	driver.part, err = NewNodePart(driver.nodeLogger, driver.arwenOutputRead, driver.arwenInputWrite, driver.blockchainHook)
+	if err != nil {
+		return err
+	}
+
+	driver.continuouslyCopyArwenLogs(arwenStdout, arwenStderr)
+
+	return nil
+}
+
+func (driver *ArwenDriver) getArwenPath() (string, error) {
+	arwenPath := os.Getenv("ARWEN_PATH")
+	if fileExists(arwenPath) {
+		return arwenPath, nil
+	}
+
+	cwd, err := os.Getwd()
+	if err != nil {
+		return "", err
+	}
+
+	arwenPath = path.Join(cwd, "arwen")
+	if fileExists(arwenPath) {
+		return arwenPath, nil
+	}
+
+	return "", common.ErrArwenNotFound
+}
+
+func fileExists(filename string) bool {
+	info, err := os.Stat(filename)
+	if os.IsNotExist(err) {
+		return false
+	}
+
+	return !info.IsDir()
+}
+
+func (driver *ArwenDriver) resetPipeStreams() error {
+	closeFile(driver.arwenInputRead)
+	closeFile(driver.arwenInputWrite)
+	closeFile(driver.arwenOutputRead)
+	closeFile(driver.arwenOutputWrite)
+	closeFile(driver.arwenLogRead)
+	closeFile(driver.arwenLogWrite)
+
+	var err error
+
+	driver.arwenInputRead, driver.arwenInputWrite, err = os.Pipe()
+	if err != nil {
+		return err
+	}
+
+	driver.arwenOutputRead, driver.arwenOutputWrite, err = os.Pipe()
+	if err != nil {
+		return err
+	}
+
+	driver.arwenLogRead, driver.arwenLogWrite, err = os.Pipe()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func closeFile(file *os.File) {
+	if file != nil {
+		err := file.Close()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Cannot close file.\n")
+		}
+	}
+}
+
+func (driver *ArwenDriver) restartArwenIfNecessary() error {
+	if !driver.IsClosed() {
+		return nil
+	}
+
+	err := driver.startArwen()
+	return err
+}
+
+// IsClosed checks whether the Arwen process is closed
+func (driver *ArwenDriver) IsClosed() bool {
+	pid := driver.command.Process.Pid
+	process, err := os.FindProcess(pid)
+	if err != nil {
+		return true
+	}
+
+	err = process.Signal(syscall.Signal(0))
+	if err != nil {
+		return true
+	}
+
+	return false
+}
+
+// RunSmartContractCreate sends a deploy request to Arwen and waits for the output
+func (driver *ArwenDriver) RunSmartContractCreate(input *vmcommon.ContractCreateInput) (*vmcommon.VMOutput, error) {
+	driver.nodeLogger.Trace("RunSmartContractCreate")
+
+	err := driver.restartArwenIfNecessary()
```

Comment:
driverRestartErrors should be treated differently in smart contract processing, like abandoning the current processing of the SC transactions until the driver is back. this has to be done in the integration process.

Reply:
Fixed: I removed calls to `driver.restartArwenIfNecessary()`, made `restartArwenIfNecessary` a public function and skipped the test `TestArwenDriver_RestartsIfStopped`.

Reply:
please add a TODO to have a clean integration for this for the node. The basic idea is that the node should not wait for arwen to restart, but process other types of transactions, and only when arwen is ready should go with the normal transactions. Still thinking of how this should be handled.

Reply:
Added TODO.